### DISCUSSION TITLE
Dynamic Sitemap

### DISCUSSION THREAD

**Discussion by nayanrabiul** [withastro/roadmap#906] [Proposal]
need for a dynamic method to generate a sitemap for server-side rendering (SSR) mode. This is particularly important in scenarios like an E-commerce website where products are constantly changing. Without such a dynamic approach, it becomes challenging to keep track of what products are available in the store.

**Comment by bholmesdev** [comment]
Thanks for sharing this @nayanrabiul! I was about to open this discussion myself per discussion [on this issue](https://github.com/withastro/astro/issues/3682#issuecomment-2063877743).

It sounds like Next.js has a solution to this, where you specify the routes you want to prerender for use in the sitemap. As a developer, you could enumerate your pages and map to URLs:
```ts
// sitemap.ts
const Pages = pages?.map((page) => ({
    url: `${baseUrl}/${page.params.pageSlug}`,
  }));
```

Trouble is, all sitemap configuration lives in your `astro.config.mjs` today. This wouldn't have access to your runtime pages. Our RSS integration gets around this by requiring you to add an `rss.xml.ts` file to your `src/pages` for generation. Perhaps we could move sitemap to a similar model?

**Reply by nayanrabiul** [reply]
yes, i didnt foundt it online, but i try it dont know why and it works. Thnaks for answering.

**Comment by Fryuni** [comment]
Although it is not for a fully dynamic sitemap, where it can change without a rebuild, you can have a look at my [Sitemap Extensions](https://inox-tools.vercel.app/sitemap-ext) built on top of the official Sitemap integration.
It doesn't solve what you are describing exactly how you want it, but depending on your use case, it might be enough.

**Reply by bholmesdev** [reply]
Wow, this is a clever design! So if I needed to generate sitemap entries for a dynamic SSR route, I could call the function from my Astro frontmatter like so?
```astro
---
// src/pages/blog/[...slug].astro
import { getCollection } from 'astro:content';
import sitemap from 'sitemap-ext:config';

sitemap(async ({ addToSitemap, removeFromSitemap }) => {
  const entries = await getCollection('blog');
  for (const entry of entries) {
    addToSitemap({ slug: entry.slug });
  }
});
---
```

**Reply by Fryuni** [reply]
Exactly!

**Reply by bholmesdev** [reply]
Sweet! In that case, this is pretty compelling as a core feature of the sitemap integration. Do you have any interest in leading a small stage 2 on this @Fryuni?

**Reply by Fryuni** [reply]
Sure, I can make an RFC for that!

It wouldn't be that small 😅
Well, it would be small but not simple, or at least not the implementation.

The extension is based on an [AIK plugin I made](https://github.com/Fryuni/inox-tools/tree/main/packages/aik-route-config) (still undocumented) that injects a separate integration to modify the result from the compiler in order to know on which route the component is used and call that function for SSR pages during build-time.

Moving it to the official integration would allow it to be much simpler but still way more complicated than the current integration.

**Reply by bholmesdev** [reply]
@Fryuni Ah, good to know the lift involved. I'm also not sure the _fully_ featureset of your integration would be required to fulfill this discussion. Would making this feature SSR-only lessen the lift?

**Reply by Fryuni** [reply]
Making the sitemap itself SSR would be a redesign of the lib since it currently all runs in build time to generate a static sitemap. But it might actually be simpler though, maybe that's a great idea.

More features, more powerful, might be interesting.
